primitives/src/v1.rs

```diff
@@ -476,3 +479,169 @@ pub struct AvailableData {
 	/// The omitted validation data.
 	pub omitted_validation: OmittedValidationData,
 }
+
+/// A helper data-type for tracking validator-group rotations.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+pub struct GroupRotationInfo<N = BlockNumber> {
+	/// The block number where the session started.
+	pub session_start_block: N,
+	/// How often groups rotate. 0 means never.
+	pub group_rotation_frequency: N,
+	/// The current block number.
+	pub now: N,
+}
+
+impl GroupRotationInfo {
+	/// Returns the index of the group needed to validate the core at the given index, assuming
+	/// the given number of cores.
+	///
+	/// `core_index` should be less than `cores`.
+	pub fn group_for_core(&self, core_index: usize, cores: usize) -> usize {
+		if self.group_rotation_frequency == 0 { return core_index }
+		if cores == 0 { return 0 }
+
+		let blocks_since_start = self.now.saturating_sub(self.session_start_block);
+		let rotations = blocks_since_start / self.group_rotation_frequency;
+
+		(core_index + rotations as usize) % cores
+	}
+}
+
+impl<N: Saturating + BaseArithmetic + Copy> GroupRotationInfo<N> {
+	/// Returns the block number of the next rotation after the current block. If the current block
+	/// is 10 and the rotation frequency is 5, this should return 15.
+	///
+	/// If the group rotation frequency is 0, returns 0.
+	pub fn next_rotation_at(&self) -> N {
+		if self.group_rotation_frequency.is_zero() { return Zero::zero() }
+
+		let cycle_once = self.now + self.group_rotation_frequency;
+		cycle_once - (
+			cycle_once.saturating_sub(self.session_start_block) % self.group_rotation_frequency
+		)
+	}
+
+	/// Returns the block number of the last rotation before or including the current block. If the
+	/// current block is 10 and the rotation frequency is 5, this should return 10.
+	///
+	/// If the group rotation frequency is 0, returns 0.
+	pub fn last_rotation_at(&self) -> N {
+		if self.group_rotation_frequency.is_zero() { return Zero::zero() }
+		self.now - (
+			self.now.saturating_sub(self.session_start_block) % self.group_rotation_frequency
+		)
+	}
+}
+
+/// Information about a core which is currently occupied.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+pub struct OccupiedCore<N = BlockNumber> {
+	/// The ID of the para occupying the core.
+	pub para: Id,
```

Comment:
in this context, does para imply both parachain and parathread?

Reply:
In this module, the type is defined as `Id`, and elsewhere we import as `ParaId`. I would prefer to have a separate PR to rename `struct Id` to `struct ParaId`, as it used to be `primitives::parachain::Id`, but now is `primitives::v1::Id` which is less semantically clear.